### PR TITLE
fix: proper cursor handling for siblings nodes

### PR DESCRIPTION
**Pull request by benouat**
This PR allows to create siblings nodes within elements such as.
Actually creation was already working, but update did not at all.

```javascript
function Template(name: string) {
  T('Hello ');
  E('b');
    t(b(name));
  e();
  T('!');
}
```

To have such use cases working, both TextNode and ElementNode need to consume properly the `cursor` even if they don't do anything (such as an Element with no bindings, or a static TextNode).

~I also added a test to validate that TextNode properly update with bindings to `null` or `undefined`.
For that purpose I had to tweak `bind()` to let it return stringified value (used for TextNode) or keep `null`(used on attributes binding in order to remove attribute if null).~

/cc @mlaval

### CODE REVIEW

File: test/iv/integration.spec.ts [hunk 1]
```diff
@@ -38,12 +38,53 @@ function renderToHtml(template: Template<any>, ctx: any) {
...
+        e();
+      }
+      expect(renderToHtml(Template, 'Hello')).toEqual('<span>Hello</span>');
+      expect(renderToHtml(Template, null)).toEqual('<span></span>');
+      expect(renderToHtml(Template, 'Hello')).toEqual('<span>Hello</span>');
+      expect(renderToHtml(Template, undefined)).toEqual('<span></span>');
+    });
+
+    it('should update mix of static/bound siblings', () => {
+      /* This usecase should never occur, as compiler will never generate such a thing
```
Comment (pkozlowski-opensource) [inline test/iv/integration.spec.ts]: Well, actually we might want to change strategy at some point and generate sibling text nodes. So for me this test is valid and we could / should remove this comment.

File: test/iv/integration.spec.ts [hunk 2]
```diff
@@ -38,12 +38,53 @@ function renderToHtml(template: Template<any>, ctx: any) {
...
+          T('Hello ');
+          t(b(name));
+          T('!');
+        e();
+      }
+      expect(renderToHtml(Template, 'world')).toEqual('<span>Hello world!</span>');
+      expect(renderToHtml(Template, 'monde')).toEqual('<span>Hello monde!</span>');
+    });
+
+    it('should support elements as surrounding siblings', () => {
```
Comment (pkozlowski-opensource) [inline test/iv/integration.spec.ts]: I'm not sure what you wanted to say here... Did you mean "should support _text_ elements as surrounding siblings"? 

Could you try to re-word name of this test to make the intention clear?

File: test/iv/integration.spec.ts [hunk 3]
```diff
@@ -38,12 +38,53 @@ function renderToHtml(template: Template<any>, ctx: any) {
...
         E('h1');
-          t(`Hello, ${name}!`);
+          t(b1('Hello, ', name, '!'));
         e();
       }
     });
   });
 
+  describe('text node bindings', () => {
+    it('should update properly to "null" or "undefined"', () => {
```
Comment (pkozlowski-opensource) [inline test/iv/integration.spec.ts]: 'should convert "null" / "undefined" to an empty string'?

File: src/iv/instructions.ts [hunk 4]
```diff
@@ -331,7 +327,7 @@ export function bind(value: any): any | NO_CHANGE {
     }
     bindingIndex++;
   }
-  return different ? value : NO_CHANGE;
+  return different ? (asString ? stringify(value) : value) : NO_CHANGE;
```
Comment (pkozlowski-opensource) [inline src/iv/instructions.ts]: This looks fishy... Couldn't we stringify where the value is actually used?

File: test/iv/integration.spec.ts [hunk 5]
```diff
@@ -38,12 +38,53 @@ function renderToHtml(template: Template<any>, ctx: any) {
 
       function Template(name: string) {
         E('h1');
-          t(`Hello, ${name}!`);
+          t(b1('Hello, ', name, '!'));
```
Comment (pkozlowski-opensource) [inline test/iv/integration.spec.ts]: 👍

File: test/iv/integration.spec.ts [hunk 6]
```diff
@@ -38,12 +38,53 @@ function renderToHtml(template: Template<any>, ctx: any) {
...
         e();
       }
     });
   });
 
+  describe('text node bindings', () => {
+    it('should update properly to "null" or "undefined"', () => {
+      function Template(name: string | null | undefined) {
+        E('span');
+          t(b(name, true));
```
Comment (pkozlowski-opensource) [inline test/iv/integration.spec.ts]: yeh, we definitively don't want to be passing `true` here as it would make the generated code bigger. Please move stringification logic to the bound text node instruction.

**Comment by benouat** [comment]
After discussing with @pkozlowski-opensource, it seems that my PR is trying to solve 2 things here.

- update of bound text node to undefined with usage of `t(b(exp))` is not working
- siblings (not only text nodes) are broken

I'll open a new PR for the update thing, and keep that one for the siblings issue.

**Comment by pkozlowski-opensource** [comment]
> update of bound text node to undefined with usage of t(b(exp)) is not working

@mhevery what was your intention for text nodes that have only one interpolation (ex. `<div>{{exp}}</div>`? It could be one of:
* `t(b(exp))`
* `t(b1('', exp, ''))`

If we assume that compiler should generate `t(b1('', exp, ''))` that stringification would work just fine, right? The downside is that it generates slightly more code.

**Comment by benouat** [comment]
> update of bound text node to undefined with usage of t(b(exp)) is not working

#19 opened to fix the update part.

Now I'll add a lot more tests here to assess that all siblings are updated correctly

**Comment by mhevery** [comment]
cleaned up and merged as 1041863
